### PR TITLE
Enable dynamic resolution for vivit

### LINKED ISSUE #30579: Community contribution: enable dynamic resolution input for more vision models.
### Feature request

Some of our models interpolate its positional embeddings, enabling pretrained checkpoints to be used on different input resolutions. For example, [here in ViT](https://github.com/huggingface/transformers/blob/75bbfd5b2237b7e35a9265731ecf63022579e7e2/src/transformers/models/vit/modeling_vit.py#L79). 

- [ ] [altclip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/altclip/modeling_altclip.py)
- [ ] [beit](https://github.com/huggingface/transformers/blob/main/src/transformers/models/beit/modeling_beit.py)
- [ ] [blip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/blip/modeling_blip.py)
- [ ] [blip_2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/blip_2/modeling_blip_2.py)
- [ ] [bridgetower](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bridgetower/modeling_bridgetower.py)
- [ ] [chinese_clip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/chinese_clip/modeling_chinese_clip.py)
- [ ] [clip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/clip/modeling_clip.py)
- [ ] [conditional_detr](https://github.com/huggingface/transformers/blob/main/src/transformers/models/conditional_detr/modeling_conditional_detr.py)
- [ ] [data2vec](https://github.com/huggingface/transformers/blob/main/src/transformers/models/data2vec/modeling_data2vec_vision.py)
- [ ] [deformable_detr](https://github.com/huggingface/transformers/blob/main/src/transformers/models/deformable_detr/modeling_deformable_detr.py)
- [ ] [deit](https://github.com/huggingface/transformers/blob/main/src/transformers/models/deit/modeling_deit.py)
- [ ] [detr](https://github.com/huggingface/transformers/blob/main/src/transformers/models/detr/modeling_detr.py)
- [ ] [donut](https://github.com/huggingface/transformers/blob/main/src/transformers/models/donut/modeling_donut_swin.py)
- [ ] [git](https://github.com/huggingface/transformers/blob/main/src/transformers/models/git/modeling_git.py)
- [ ] [grounding_dino](https://github.com/huggingface/transformers/blob/main/src/transformers/models/grounding_dino/modeling_grounding_dino.py)
- [ ] [idefics2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/idefics2/modeling_idefics2.py)
- [ ] [kosmos2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/kosmos2/modeling_kosmos2.py)
- [ ] [maskformer swin](https://github.com/huggingface/transformers/blob/main/src/transformers/models/maskformer/modeling_maskformer_swin.py)
- [ ] [owlv2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/owlv2/modeling_owlv2.py)
- [ ] [owlvit](https://github.com/huggingface/transformers/blob/main/src/transformers/models/owlvit/modeling_owlvit.py)
- [ ] [perceiver](https://github.com/huggingface/transformers/blob/main/src/transformers/models/perceiver/modeling_perceiver.py)
- [ ] [siglip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/siglip/modeling_siglip.py)
- [ ] [swin](https://github.com/huggingface/transformers/blob/main/src/transformers/models/swin/modeling_swin.py)
- [ ] [swin2sr](https://github.com/huggingface/transformers/blob/main/src/transformers/models/swin2sr/modeling_swin2sr.py)
- [ ] [swinv2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/swinv2/modeling_swinv2.py)
- [ ] [table_transformer](https://github.com/huggingface/transformers/blob/main/src/transformers/models/table_transformer/modeling_table_transformer.py)
- [ ] [tvp](https://github.com/huggingface/transformers/blob/main/src/transformers/models/tvp/modeling_tvp.py)
- [ ] [vit_mae](https://github.com/huggingface/transformers/blob/main/src/transformers/models/vit_mae/modeling_vit_mae.py)
- [ ] [vivit](https://github.com/huggingface/transformers/blob/main/src/transformers/models/vivit/modeling_vivit.py)

### Motivation

Let's add this to more models, to leverage existing checkpoints for new cases! 

### Your contribution

For anyone who would like to contribute, please comment on the issue, claiming a model you'd like to work on and share a link to the PR. 

Each PR should: 
* Add an `interpolate_pos_encoding` method
* Add a test showing the model can correctly interpolate an input image of a different size

There was a PR opened to add this to CLIP models, which is now inactive, but useful for reference of the changes to make: https://github.com/huggingface/transformers/pull/27457

Once the PR is ready, you can ping me for review 🤗

### PR DESCRIPTION
**Pull request by jla524**
# What does this PR do?

Addresses #30579 (issue)

## Who can review?

@amyeroberts

### CODE REVIEW

File: tests/models/vivit/test_modeling_vivit.py [hunk 1]
```diff
@@ -353,3 +353,31 @@ def test_inference_for_video_classification(self):
...
+            outputs = model(pixel_values, interpolate_pos_encoding=True)
+
+        # verify the logits
+        expected_shape = torch.Size((1, 3137, 768))
+        self.assertEqual(outputs.last_hidden_state.shape, expected_shape)
+
+        expected_slice = torch.tensor(
+            [[-0.5608, 0.6402, 0.1368], [-0.1167, -0.1180, -0.3220], [-0.1988, -0.2729, -0.0176]]
+        ).to(torch_device)
+        print(outputs.last_hidden_state[0, :3, :3])
```
Comment (nileshkokane01) [inline tests/models/vivit/test_modeling_vivit.py:381]: Should not have print statement here.

File: src/transformers/models/vivit/modeling_vivit.py [hunk 2]
```diff
@@ -67,12 +67,14 @@ def __init__(self, config):
...
-        if height != self.image_size or width != self.image_size:
-            raise ValueError(
-                f"Input image size ({height}*{width}) doesn't match model ({self.image_size}*{self.image_size})."
-            )
+        if not interpolate_pos_encoding:
+            if height != self.image_size or width != self.image_size:
+                raise ValueError(
+                    f"Image image size ({height}*{width}) doesn't match model"
+                    f" ({self.image_size[0]}*{self.image_size[1]})."
+                )
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:77]:
```suggestion
        if not interpolate_pos_encoding and (height != self.image_size or width != self.image_size):
            raise ValueError(
                f"Image image size ({height}*{width}) doesn't match model"
                f" ({self.image_size[0]}*{self.image_size[1]})."
            )
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 3]
```diff
@@ -102,16 +104,52 @@ def __init__(self, config):
...
+        h0, w0 = h0 + 0.1, w0 + 0.1
+        patch_pos_embed = patch_pos_embed.reshape(1, int(math.sqrt(num_positions)), int(math.sqrt(num_positions)), dim)
+        patch_pos_embed = patch_pos_embed.permute(0, 3, 1, 2)
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed,
+            scale_factor=(h0 / math.sqrt(num_positions), w0 / math.sqrt(num_positions)),
+            mode="bicubic",
+            align_corners=False,
+        )
+        assert int(h0) == patch_pos_embed.shape[-2] and int(w0) == patch_pos_embed.shape[-1]
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:136]: nit - I realise this comes from vit. The convention in the library is not to have assert statements in the modeling files 

```suggestion
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 4]
```diff
@@ -102,16 +104,52 @@ def __init__(self, config):
...
+        patch_pos_embed = patch_pos_embed.permute(0, 2, 3, 1).view(1, -1, dim)
+        return torch.cat((class_pos_embed.unsqueeze(0), patch_pos_embed), dim=1)
 
+    def forward(self, pixel_values, interpolate_pos_encoding=interpolate_pos_encoding):
+        print(pixel_values.shape)
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:141]:
```suggestion
```

File: tests/models/vivit/test_modeling_vivit.py [hunk 5]
```diff
@@ -353,3 +353,30 @@ def test_inference_for_video_classification(self):
...
+
+        # verify the logits
+        expected_shape = torch.Size((1, 3137, 768))
+        self.assertEqual(outputs.last_hidden_state.shape, expected_shape)
+
+        expected_slice = torch.tensor(
+            [[-0.5608, 0.6402, 0.1368], [-0.1167, -0.1180, -0.3220], [-0.1988, -0.2729, -0.0176]]
+        ).to(torch_device)
+
+        self.assertTrue(torch.allclose(outputs.last_hidden_state[0, :3, :3], expected_slice, atol=1e-4))
```
Comment (amyeroberts) [inline tests/models/vivit/test_modeling_vivit.py:382]: No need to test the logits unless they're taken from the original model for this image size 

```suggestion
```

File: tests/models/vivit/test_modeling_vivit.py [hunk 6]
```diff
@@ -353,3 +353,30 @@ def test_inference_for_video_classification(self):
...
+        # Vivit models have an `interpolate_pos_encoding` argument in their forward method,
+        # allowing to interpolate the pre-trained position embeddings in order to use
+        # the model on higher resolutions. The DINO model by Facebook AI leverages this
+        # to visualize self-attention on higher resolution images.
+        model = VivitModel.from_pretrained("google/vivit-b-16x2").to(torch_device)
+
+        image_processor = VivitImageProcessor.from_pretrained("google/vivit-b-16x2", size=480)
+        video = prepare_video()
+        inputs = image_processor(video, return_tensors="pt")
+        pixel_values = inputs.pixel_values.to(torch_device)
```
Comment (amyeroberts) [inline tests/models/vivit/test_modeling_vivit.py:368]: * We should use the image processor already created
* image processor's behaviour can be configured when they're called 
* size should be a dict 
* I believe the value we want to modify for this is also `crop_size` 

```suggestion
        image_processor = VivitImageProcessor.from_pretrained("google/vivit-b-16x2")
        video = prepare_video()
        inputs = image_processor(video, size={"shortest_edge": 480}, crop_size={"height": 480, "width": 480}, return_tensors="pt")
        pixel_values = inputs.pixel_values.to(torch_device)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Just a few nit comments

File: tests/models/vivit/test_modeling_vivit.py (hunk 6 above)
Comment (jla524) [reply tests/models/vivit/test_modeling_vivit.py:368]: should we use `image_processor = self.default_image_processor` instead?

File: tests/models/vivit/test_modeling_vivit.py [hunk 7]
```diff
@@ -353,3 +353,24 @@ def test_inference_for_video_classification(self):
...
+        image_processor = VivitImageProcessor.from_pretrained("google/vivit-b-16x2")
+        video = prepare_video()
+        inputs = image_processor(video, size={"shortest_edge": 480}, crop_size={"height": 480, "width": 480}, return_tensors="pt")
+        pixel_values = inputs.pixel_values.to(torch_device)
+
+        # forward pass
+        with torch.no_grad():
+            outputs = model(pixel_values, interpolate_pos_encoding=True)
+
+        # verify the logits
```
Comment (amyeroberts) [inline tests/models/vivit/test_modeling_vivit.py:374]:
```suggestion
        # verify the logits shape
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 8]
```diff
@@ -67,11 +67,12 @@ def __init__(self, config):
             config.num_channels, config.hidden_size, kernel_size=config.tubelet_size, stride=config.tubelet_size
         )
 
-    def forward(self, pixel_values):
+    def forward(self, pixel_values, interpolate_pos_encoding=False):
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:70]:
```suggestion
    def forward(self, pixel_values, interpolate_pos_encoding: bool = False):
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 9]
```diff
@@ -102,16 +103,50 @@ def __init__(self, config):
...
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed,
+            scale_factor=(h0 / math.sqrt(num_positions), w0 / math.sqrt(num_positions)),
+            mode="bicubic",
+            align_corners=False,
+        )
+        patch_pos_embed = patch_pos_embed.permute(0, 2, 3, 1).view(1, -1, dim)
+        return torch.cat((class_pos_embed.unsqueeze(0), patch_pos_embed), dim=1)
+
+    def forward(self, pixel_values, interpolate_pos_encoding=interpolate_pos_encoding):
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:138]: It's not possible to define a default like this unless interpolate_pos_encoding is a global

```suggestion
    def forward(self, pixel_values, interpolate_pos_encoding: bool = False):
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 10]
```diff
@@ -437,6 +472,8 @@ def _init_weights(self, module):
         output_hidden_states (`bool`, *optional*):
             Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
             more detail.
+        interpolate_pos_encoding (`bool`, *optional*):
+            Whether to interpolate the pre-trained position encodings.
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:476]: I realise this comes from vit - but let's default to `False` here. It's easier than having to deal with null/true/false 

```suggestion
        interpolate_pos_encoding (`bool`, *optional*, `False`):
            Whether to interpolate the pre-trained position encodings.
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 11]
```diff
@@ -482,6 +519,7 @@ def forward(
         head_mask: Optional[torch.FloatTensor] = None,
         output_attentions: Optional[bool] = None,
         output_hidden_states: Optional[bool] = None,
+        interpolate_pos_encoding: Optional[bool] = None,
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:522]:
```suggestion
        interpolate_pos_encoding: bool = False,
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 12]
```diff
@@ -622,6 +670,7 @@ def forward(
         labels: Optional[torch.LongTensor] = None,
         output_attentions: Optional[bool] = None,
         output_hidden_states: Optional[bool] = None,
+        interpolate_pos_encoding: Optional[bool] = None,
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:673]:
```suggestion
        interpolate_pos_encoding: bool = False,
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for improving the models and iterating on this! 

Just a few small nits. For the quality checks, running `make fixup` and pushing the changes should resolve them

**Review by amyeroberts — APPROVED** [review]
Thanks for working on this and improving the library's models!
